x/farming/keeper/genesis.go

```diff
@@ -16,68 +16,66 @@ func (k Keeper) InitGenesis(ctx sdk.Context, genState types.GenesisState) {
 	moduleAcc := k.accountKeeper.GetModuleAccount(ctx, types.ModuleName)
 	k.accountKeeper.SetModuleAccount(ctx, moduleAcc)
 
-	// TODO: decision making is needed
-	// feeCollectorAcc, err := sdk.AccAddressFromBech32(types.DefaultFarmingFeeCollector)
-	// if err != nil {
-	// 	panic(err)
-	// }
-
-	// k.accountKeeper.SetModuleAccount(ctx, feeCollectorAcc)
-
-	// TODO: unimplemented
-	//for _, record := range genState.PlanRecords {
-	//	k.SetPlanRecord(ctx, record)
-	//}
-	//for _, staking := range genState.Stakings {
-	//	k.SetStaking(ctx, staking)
-	//}
-	//for _, reward := range genState.Rewards {
-	//	k.SetReward(ctx, reword)
-	//}
+	_, err := sdk.AccAddressFromBech32(genState.Params.FarmingFeeCollector)
+	if err != nil {
+		panic(err)
+	}
 
+	for _, record := range genState.PlanRecords {
+		plan, err := types.UnpackPlan(&record.Plan)
+		if err != nil {
+			panic(err)
+		}
+		k.SetPlan(ctx, plan)
+		k.SetGlobalPlanId(ctx, plan.GetId())
+	}
+	for _, staking := range genState.Stakings {
+		k.SetStaking(ctx, staking)
+	}
+	for _, reward := range genState.Rewards {
+		k.SetReward(ctx, reward.StakingCoinDenom, reward.GetFarmer(), reward.RewardCoins)
+	}
 }
 
 // ExportGenesis returns the farming module's genesis state.
 func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
 	params := k.GetParams(ctx)
-
-	// TODO: unimplemented
 	var planRecords []types.PlanRecord
 
-	//plans := k.GetAllPlans(ctx)
-	//stakings := k.GetAllStakings(ctx)
-	//rewards := k.GetAllRewards(ctx)
-
-	//for _, plan := range plans {
-	//	record, found := k.GetPlanRecord(ctx, plan)
-	//	if found {
-	//		planRecords = append(planRecords, record)
-	//	}
-	//}
-	//
-	//if len(planRecords) == 0 {
-	//	planRecords = []types.PlanRecord{}
-	//}
+	plans := k.GetAllPlans(ctx)
+	stakings := k.GetAllStakings(ctx)
+	rewards := k.GetAllRewards(ctx)
+
+	for _, plan := range plans {
+		any, err := types.PackPlan(plan)
+		if err != nil {
+			panic(err)
+		}
+		planRecords = append(planRecords, types.PlanRecord{
+			Plan:             *any,
+			FarmingPoolCoins: k.bankKeeper.GetAllBalances(ctx, plan.GetFarmingPoolAddress()),
+		})
+	}
 
-	return types.NewGenesisState(params, planRecords, nil, nil)
+	epochTime, _ := k.GetLastEpochTime(ctx)
```

Comment:
`LastEpochTime` may be not present on the store so the second return value `found` can be `false`. In that case, calling `NewGenesisState` with concrete `time.Time` value will not act as expected.

Reply:
That's right. If there's no value(false), the expectation is to return `time.Time{}` and set it to this empty `time.Time{}` value for export genesis. Can't we do this?